In [9]:
import pandas as pd
import numpy as np

In [244]:
df = pd.read_csv('train.csv')

In [245]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [246]:
#(df['Survived'].value_counts()/df['Survived'].value_counts().sum()*100).round(2)

In [247]:
def is_adult(age):
   if age >= 18:
       return 'adult'
   else:
       return 'kid'

In [248]:
df.dropna(subset=['Age'],inplace=True)

In [249]:
df.shape

(714, 12)

In [250]:
df['Age'].isnull().value_counts()

False    714
Name: Age, dtype: int64

In [251]:
df['is_adult']=df['Age'].apply(is_adult)

In [252]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked is_adult  
0      0         A/5 21171   7.2500   NaN        S    adult  
1      0          PC 17599  71.2833   C85        C    adult  
2      0  STON/O2. 3101282   7.9250   NaN        S    adult  
3      0            113803  53.1000  C123        S    adult  
4      0            373450   8.0500   NaN        S    adult

#### Percentage of survival  & non survival ############

In [253]:
(df['Survived'].value_counts()/df['Survived'].value_counts().sum()*100).round(2)

0    59.38
1    40.62
Name: Survived, dtype: float64

####### As Survived Column is in binary. finding mean of Survived column will also give us Survival percentage. we see same percentage value #####

In [254]:
(df['Survived'].mean()*100)

40.61624649859944

--------------------------------------------------------------------------------------------------------------------

##### Percentage of male kid & Female kid survived out of total population ####

In [169]:
df[(df['Survived']==1) & (df['is_adult']=='kid')]['Sex'].value_counts()/df.shape[0]*100

female    5.322129
male      3.221289
Name: Sex, dtype: float64

##### Percentage of male kid & Female adult survived out of total population ####

In [170]:
df[(df['Survived']==1) & (df['is_adult']=='adult')]['Sex'].value_counts()/df.shape[0]*100

female    22.268908
male       9.803922
Name: Sex, dtype: float64

### Summing all these percentage will give you around 40.6 which is overall survived persentage as seen in 1st question####

----------------------------------------------------------------------------------------------------------

###### Percentage of male/female survived wrt adult/kid category. Here denominator is not total population but adult survival population                      so out of Total Adult Survived what is the distribution of male/female ########


In [ ]:
### percentage of male/female w.r.t Adult Survived #######

In [118]:
df[(df['Survived']==1) & (df['is_adult']=='adult')]['Sex'].value_counts()/df[(df['Survived']==1) & (df['is_adult']=='adult')]['Sex'].value_counts().sum()*100

female    69.432314
male      30.567686
Name: Sex, dtype: float64

In [ ]:
#### Kid Survived ######### out of Total kid survived what is distribution of male/female

In [119]:
df[(df['Survived']==1) & (df['is_adult']=='kid')]['Sex'].value_counts()/df[(df['Survived']==1) & (df['is_adult']=='kid')]['Sex'].value_counts().sum()*100

female    62.295082
male      37.704918
Name: Sex, dtype: float64

---------------------------------------------------------------------------------------------------

### Percentage of passengers per embark ########

In [255]:
df.groupby('Embarked').count()

PassengerId  Survived  Pclass  Name  Sex  Age  SibSp  Parch  Ticket  \
Embarked                                                                        
C                 130       130     130   130  130  130    130    130     130   
Q                  28        28      28    28   28   28     28     28      28   
S                 554       554     554   554  554  554    554    554     554   

          Fare  Cabin  is_adult  
Embarked                         
C          130     65       130  
Q           28      2        28  
S          554    116       554

In [185]:
df.groupby('Embarked').count()['Age']/df.shape[0]*100

Embarked
C    18.207283
Q     3.921569
S    77.591036
Name: Age, dtype: float64

-------------------------------------------------------------------------------------

##### Survival Ratio per Embarked ##########

In [208]:
df.groupby(['Embarked','Survived'])['PassengerId'].count()

Embarked  Survived
C         0            51
          1            79
Q         0            20
          1             8
S         0           353
          1           201
Name: PassengerId, dtype: int64

In [209]:
df.groupby(['Embarked'])['PassengerId'].count()

Embarked
C    130
Q     28
S    554
Name: PassengerId, dtype: int64

In [210]:
df.groupby(['Embarked','Survived'])['PassengerId'].count()/df.groupby(['Embarked'])['PassengerId'].count()*100

Embarked  Survived
C         0           39.230769
          1           60.769231
Q         0           71.428571
          1           28.571429
S         0           63.718412
          1           36.281588
Name: PassengerId, dtype: float64

###### Above Survival percentage can be obtained by calculating mean of Survived column after groupby 'Embarked' #####

In [136]:
df.groupby('Embarked')['Survived'].mean()*100

Embarked
C    60.769231
Q    28.571429
S    36.281588
Name: Survived, dtype: float64

---------------------------------------------------------------------------------

###### Survival ratio of cabin vs non-cabin passengers #######

In [220]:
 ########### survived percentage of people without Cabin ########

In [218]:
df[df['Cabin'].isnull()]['Survived'].mean()*100

31.190926275992435

In [ ]:
 ########### survived percentage of people with Cabin ########

In [219]:
df[df['Cabin'].notnull()]['Survived'].mean()*100

67.56756756756756

--------------------------------------------------------------------------------------------

##### Is there correlation between ticket fare and survival ##########

In [256]:
df['fare_group']=pd.qcut(df['Fare'],10,range(1,11))

In [257]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked is_adult fare_group  
0      0         A/5 21171   7.2500   NaN        S    adult          1  
1      0          PC 17599  71.2833   C85        C    adult          9  
2      0  STON/O2. 3101282   7.9250   NaN        S    adult          3  
3      0            113803  53.1000  C123        S    adult          9  
4      0            373450   8.0500   NaN        S    adult          3

In [265]:
df.groupby('fare_group')['Survived'].mean()

fare_group
1     0.192771
2     0.150000
3     0.239437
4     0.375000
5     0.312500
6     0.489796
7     0.458333
8     0.382353
9     0.628571
10    0.777778
Name: Survived, dtype: float64

In [267]:
temp = pd.DataFrame(df.groupby('fare_group')['Survived'].mean())

In [268]:
temp

Survived
fare_group          
1           0.192771
2           0.150000
3           0.239437
4           0.375000
5           0.312500
6           0.489796
7           0.458333
8           0.382353
9           0.628571
10          0.777778

In [269]:
temp.reset_index(inplace=True)

In [270]:
temp

fare_group  Survived
0          1  0.192771
1          2  0.150000
2          3  0.239437
3          4  0.375000
4          5  0.312500
5          6  0.489796
6          7  0.458333
7          8  0.382353
8          9  0.628571
9         10  0.777778

In [156]:
temp['fare_group'].corr(temp['Survived'])

0.912784046954763

##### Pearson correlation co-efficient of 0.91 shows good positive relation between Fare Price & Survival 